solver-next/internal/pipe/pipe.go

```diff
@@ -0,0 +1,198 @@
+package pipe
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+)
+
+type channel struct {
+	OnSendCompletion func()
+	mu               sync.Mutex
+	value            atomic.Value
+	lastValue        interface{}
+}
+
+func (c *channel) Send(v interface{}) {
+	c.value.Store(v)
+	if c.OnSendCompletion != nil {
+		c.OnSendCompletion()
+	}
+}
+
+func (c *channel) Receive() (interface{}, bool) {
+	v := c.value.Load()
+	if c.lastValue == v {
+		return nil, false
+	}
+	c.lastValue = v
+	return v, true
+}
+
+type Pipe struct {
+	Sender              Sender
+	Receiver            Receiver
+	OnReceiveCompletion func()
+	OnSendCompletion    func()
+}
+
+type Request struct {
+	Payload  interface{}
+	Canceled bool
+}
+
+type Sender interface {
+	Request() Request
+	Update(v interface{})
+	Finalize(v interface{}, err error)
+	Status() Status
+}
+
+type Receiver interface {
+	Receive() bool
```

Comment:
Is there a specific reason not to return `(bool, Status)`?

Reply:
The client needs access to the last state in many places, not only to detect the change event. So with current usage, the status value would always be ignored by the client.